Detect status before enable/disable addon

fenglixa opened the pull request:
- Fix #4417

- UT Result PASS:
![image](https://user-images.githubusercontent.com/12505594/58941022-84e1e700-87ad-11e9-8b65-024585e14f43.png)

- Build & Test PASS:
![image](https://user-images.githubusercontent.com/12505594/58941077-9e832e80-87ad-11e9-8a19-3f5133979699.png)

RA489 commented:
@minikube-bot ok-to-test

medyagh commented:
@minikube-bot ok-to-test

hpandeycodeit commented on cmd/minikube/cmd/config/util.go:
```diff
@@ -138,7 +139,17 @@ func EnableOrDisableAddon(name string, val string) error {
 func enableOrDisableAddonInternal(addon *assets.Addon, cmd bootstrapper.CommandRunner, data interface{}, enable bool) error {
 	var err error
+	// check addon status before enable/disable
+	ok, err := addon.IsEnabled()
```
Can you change this variable name to be little more informative?

fenglixa replied:
Address done. Thanks

hpandeycodeit commented on cmd/minikube/cmd/config/util.go:
```diff
@@ -138,7 +139,17 @@ func EnableOrDisableAddon(name string, val string) error {
+	}
+
 	if enable {
+		if ok {
```
I think this logic can be moved to the respective command files.

fenglixa replied:
Address done. Thanks

hpandeycodeit commented on cmd/minikube/cmd/config/util.go:
```diff
@@ -155,6 +166,10 @@ func enableOrDisableAddonInternal(addon *assets.Addon, cmd bootstrapper.CommandR
 		}
 	} else {
+		if !ok {
+			console.ErrStyle(console.Conflict, "addon %s is already disabled.", addon.GetAddon())
```
same for this too.

fenglixa replied:
Address done. Thanks
Append my new UT result:
![image](https://user-images.githubusercontent.com/12505594/59002746-084f1700-8846-11e9-8615-459e1a915f0f.png)

fenglixa commented:
@minikube-bot OK to test

medyagh commented:
Thanks for contribution, could you pleas add unit testint ? Maybe have a mock enabled addon and try to enable it

fenglixa commented:
> Thanks for contribution, could you pleas add unit testint ? Maybe have a mock enabled addon and try to enable it

Thanks @medyagh  Did you mean following? I think I appended it already. : )

```
 > ~/Documents/GitHub/minikube >> minikube addons list |grep enable
- addon-manager: enabled
- dashboard: enabled
- default-storageclass: enabled
- storage-provisioner: enabled
 ~/Documents/GitHub/minikube>> minikube addons enable dashboard
💥  addon dashboard was already enabled. 
```

medyagh commented:
> > > Thanks for contribution, could you pleas add unit testint ? Maybe have a mock enabled addon and try to enable it
> > 
> > 
> > Thanks @medyagh Did you mean following? I think I appended it already. : )
> > ```
> >  > ~/Documents/GitHub/minikube >> minikube addons list |grep enable
> > - addon-manager: enabled
> > - dashboard: enabled
> > - default-storageclass: enabled
> > - storage-provisioner: enabled
> >  ~/Documents/GitHub/minikube>> minikube addons enable dashboard
> > 💥  addon dashboard was already enabled. 
> > ```
> 
> I meant like for example for this logic in your code
> 
> ```
> 	if addonStatus && enable {
> 		console.ErrStyle(console.Conflict, "addon %s was already enabled.", addon.GetAddon())
> 		os.Exit(exit.Unavailable)
> 	} else if !addonStatus && !enable {
> 		console.ErrStyle(console.Conflict, "addon %s was already disabled.", addon.GetAddon())
> 		os.Exit(exit.Unavailable)
> 	}
> ```
> 
> you could put it in a func and then add a TestFunc for it.
> that you could do unit testing for that func, here are some example unit tests in minikube. https://github.com/kubernetes/minikube/blob/master/cmd/minikube/cmd/config/util_test.go#L69
> 
> you could have a unit test that tries to disable an already disabled addon, and expects it to get err

for example you could have a func called

`func isAddonAlreadySet(addon *assets.Addon, enable bool) bool`
and then you can move the logic of  `addonStatus, err := addon.IsEnabled()`  inside that func

sharifelgamal commented:
@minikube-bot OK to test

fenglixa commented:
@medyagh Thanks for your comments. I added unit test cases for the new added func. Please help check & review.
@sharifelgamal I addressed the issue in the result of integration test, please help check again. Thanks.
(the reason is get dashboard url will enable dashboard always, which will be failed as it was already enabled)

BR
-Francis

fenglixa commented on cmd/minikube/cmd/config/util.go:
```diff
@@ -138,6 +139,21 @@ func EnableOrDisableAddon(name string, val string) error {
+	}
+
 	if enable {
 		for _, addon := range addon.Assets {
```
removed. Thanks

fenglixa commented:
```
09:34:38 | >   - error: failed to create listener: failed to listen on 0.0.0.0:8443: listen tcp 0.0.0.0:8443: bind: address already in use
--- FAIL: TestVersionUpgrade (127.96s)
```
should be case issue related to #4418

fenglixa commented:
@minikube-bot OK to test

RA489 commented:
/ok-to-test

medyagh commented on cmd/minikube/cmd/config/util.go:
```diff
@@ -137,20 +137,28 @@ func EnableOrDisableAddon(name string, val string) error {
 	if addonStatus && enable {
-		console.ErrStyle(console.Conflict, "addon %s was already enabled.", addon.GetAddon())
-		os.Exit(exit.Unavailable)
+		return fmt.Errorf("addon %s was already enabled", addon.GetAddon())
```
for more consistency, lets use errors.Wrap in stead of `fmt.Errorf`

fenglixa replied:
Thanks Medya, should not errors.Wrap for this case

medyagh commented on cmd/minikube/cmd/config/util_test.go:
```diff
@@ -79,3 +80,15 @@ func TestSetBool(t *testing.T) {
+		addonStatus, _ := addon.IsEnabled()
+		err := isAddonAlreadySet(addon, addonStatus)
+		if err == nil {
+			t.Fatalf("did not return error for the addon was already set")
```
Thanks for adding the unit tests, two things to impriove this unit test
1- first lets change the t.Fatalf to t.Error
2- lets make the message in the following format:
```
t.Errorf("Expected to get error , got nil", )
```

fenglixa replied:
Done. Thanks

medyagh commented on pkg/minikube/assets/addons.go:
```diff
@@ -47,6 +47,11 @@ func NewAddon(assets []*BinAsset, enabled bool, addonName string) *Addon {
 }
 
+// GetAddon get the addon name
+func (a *Addon) GetAddon() string {
```
lets change this to Name()
since this does not return the addon but returns a string.

fenglixa replied:
Done. Thanks Medya

medyagh replied:
in go style usually there is no need to say Get, and also since the func on an addon, there is no need to say Addon in the name either, so I suggest changing it to just Name()

fenglixa replied:
Thanks Medya, I changed it to Name().

medyagh commented on cmd/minikube/cmd/config/util.go:
```diff
@@ -136,8 +137,31 @@ func EnableOrDisableAddon(name string, val string) error {
+	// check addon status before enabling/disabling it
+	if err := isAddonAlreadySet(addon, enable); err != nil {
+		console.ErrStyle(console.Conflict, "%v", err)
+		os.Exit(exit.Unavailable)
```
lets exit with 0, because we don't want to fail automated minikubes tests used CI, if their addon was already set, this will make their script to fail.

I rather we just warn the user but not exit with error. lets exit with code 0

fenglixa replied:
Done. Thanks Medya

fenglixa commented:
/retest

fenglixa commented:
/retest

fenglixa commented:
Test failed is not related to my modification.

/retest

medyagh commented on pkg/minikube/assets/addons.go:
```diff
@@ -47,6 +47,11 @@ func NewAddon(assets []*BinAsset, enabled bool, addonName string) *Addon {
 }
 
+// GetAddonName get the addon name
+func (a *Addon) GetAddonName() string {
```
Thanks for all the changes, please change this to Name(). I think after this change it will be good to go !

fenglixa replied:
Done. Thanks

RA489 commented:
@fenglixa can you please squash the commit to one

fenglixa commented:
@RA489  squashed done. Thanks

RA489 commented:
fenglixa Thanks!

fenglixa commented:
Seems it can not take effect when I provide /retest command. Could someone help trigger retest of integration? Thanks

RA489 commented:
/retest

fenglixa commented:
Thanks @RA489.  but seems it is also not take effect.

sharifelgamal commented:
@minikube-bot OK to test

medyagh commented:
@minikube-bot okay to test

fenglixa commented:
@minikube-bot okay to test

medyagh commented:
@minikube-bot okay to test.

medyagh commented:
@minikube-bot okay to test

medyagh commented:
@minikube-bot OK to test

medyagh commented:
@fenglixa Thank you so much for patiently making all the required changes and sorry for the unrelated failing tests. Thank you for the contribution
